File: problems/[144]二叉树的前序遍历.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        def dp(ans: List[int], root: TreeNode):
            if root:
                ans.append(root.val)
                dp(ans, root.left)
                dp(ans, root.right)
        
        ans = list()
        dp(ans, root)
        return ans

File: solutions/[144]二叉树的前序遍历.cpp
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	vector<int> preorderTraversal(TreeNode* root) {
		vector<int> ans;
		dp(root, ans);
	}
	void dp(TreeNode* root, vector<int>& ans) {
		if (root == nullptr)
			return;
		ans.push_back(root->val);
		dp(root->left, ans);
		dp(root->right, ans);
	}
};